backends/ebpf/psa/externs/ebpfPsaRegister.cpp

```diff
@@ -0,0 +1,230 @@
+                                      1, ret.c_str());
+
+    builder->blockEnd(true);
+
+    builder->blockEnd(true);
+}
+
+void EBPFRegisterPSA::emitInstance(CodeBuilder *builder) {
+    builder->target->emitTableDecl(builder, instanceName,
+                                   this->useArrayMap ? TableArray : TableHash,
+                                   this->keyTypeName,
+                                   this->valueTypeName, size);
+}
+
+void EBPFRegisterPSA::emitRegisterRead(CodeBuilder* builder, const P4::ExternMethod* method,
+                                       ControlBodyTranslatorPSA* translator,
+                                       const IR::Expression* leftExpression) {
+    auto indexArg = method->expr->arguments->at(0)->expression->to<IR::PathExpression>();
+    cstring indexParamName = translator->getActionParamName(indexArg);
+    BUG_CHECK(!indexParamName.isNullOrEmpty(), "Index param cannot be empty");
+
+    cstring valueName = program->refMap->newName("value");
+
+    builder->emitIndent();
+    this->valueType->declare(builder, valueName, true);
+    builder->endOfStatement(true);
+
+    cstring msgStr = Util::printf_format("Register: reading %s",
+                                 instanceName.c_str());
+    builder->target->emitTraceMessage(builder, msgStr.c_str());
+
+    builder->emitIndent();
+    builder->target->emitTableLookup(builder, dataMapName, indexParamName, valueName);
+    builder->endOfStatement(true);
+
+    builder->emitIndent();
+    builder->appendFormat("if (%s != NULL) ", valueName.c_str());
+    builder->blockStart();
+    builder->emitIndent();
+    if (leftExpression != nullptr) {
+        codeGen->visit(leftExpression);
+        builder->appendFormat(" = *%s", valueName);
+        builder->endOfStatement(true);
+    }
+    builder->target->emitTraceMessage(builder, "Register: Entry found!");
+    builder->blockEnd(false);
+    builder->appendFormat(" else ");
+    builder->blockStart();
+    builder->emitIndent();
+
+    if (leftExpression != nullptr) {
+        codeGen->visit(leftExpression);
+        builder->append(" = ");
+        if (this->initialValue != nullptr) {
+            builder->append(this->initialValue->value.str());
+        } else {
+            builder->append("(");
+            this->valueType->declare(builder, cstring::empty, false);
+            builder->append(")");
+            this->valueType->emitInitializer(builder);
+        }
+        builder->endOfStatement(true);
+    }
+
+    builder->target->emitTraceMessage(builder, "Register: Entry not found, using default value");
+    builder->blockEnd(true);
+}
+
+void EBPFRegisterPSA::emitRegisterWrite(CodeBuilder* builder, const P4::ExternMethod* method,
+                                        ControlBodyTranslatorPSA* translator) {
+    auto indexArgExpr = method->expr->arguments->at(0)->expression->to<IR::PathExpression>();
+    cstring indexParamName = translator->getActionParamName(indexArgExpr);
```

Comment:
is this implying that within an action the index must be an action parameter?
So it cannot be some expression that involves the action parameter?

Reply:
the function name is misleading, I'll change it

Reply:
I believe that the behavior that you mention is tested in https://github.com/p4lang/p4c/blob/1c179a11f52f2b38c7db5e3fda68e16893ed05b2/backends/ebpf/tests/p4testdata/register-action.p4.